File: HoTT-MdlViewer-JavaFX/src/main/java/gde/mdl/ui/ExceptionDialog.java
package gde.mdl.ui;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.logging.Level;
import java.util.logging.Logger;

import gde.mdl.messages.Messages;
import javafx.application.Platform;
import javafx.scene.control.Alert;
import javafx.scene.control.Label;
import javafx.scene.control.TextArea;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Priority;

public class ExceptionDialog extends Alert {
    private static final Logger LOG = Logger.getLogger(ExceptionDialog.class.getName());

    public static void show(final Throwable throwable) {
        Platform.runLater(() -> new ExceptionDialog(throwable).showAndWait());
    }

    public ExceptionDialog(final Throwable throwable) {
        super(AlertType.ERROR);

        LOG.log(Level.SEVERE, throwable.getMessage(), throwable);

        setTitle(Messages.getString("Error"));
        setHeaderText(null);

        String message = throwable.getLocalizedMessage();
        if (message == null) message = throwable.getMessage();
        if (message == null) message = throwable.getClass().getSimpleName();
        setContentText(message);

        // Create expandable Exception.
        final StringWriter sw = new StringWriter();
        final PrintWriter pw = new PrintWriter(sw);
        throwable.printStackTrace(pw);
        final String exceptionText = sw.toString();

        final Label label = new Label("The exception stacktrace was:");

        final TextArea textArea = new TextArea(exceptionText);
        textArea.setEditable(false);
        textArea.setWrapText(true);

        textArea.setMaxWidth(Double.MAX_VALUE);
        textArea.setMaxHeight(Double.MAX_VALUE);
        GridPane.setVgrow(textArea, Priority.ALWAYS);
        GridPane.setHgrow(textArea, Priority.ALWAYS);

        final GridPane expContent = new GridPane();
        expContent.setMaxWidth(Double.MAX_VALUE);
        expContent.add(label, 0, 0);
        expContent.add(textArea, 0, 1);

        // Set expandable Exception into the dialog pane.
        getDialogPane().setExpandableContent(expContent);
    }
}

File: HoTT-UI/src/main/kotlin/de/treichels/hott/ui/ExceptionDialog.kt
package de.treichels.hott.ui

import javafx.scene.control.Alert
import javafx.scene.layout.Priority
import javafx.stage.Stage
import tornadofx.*
import java.io.PrintWriter
import java.io.StringWriter
import java.util.logging.Logger

class ExceptionDialog(throwable: Throwable) : Alert(Alert.AlertType.ERROR) {
    init {
        Logger.getLogger(javaClass.name).throwing(javaClass.name, "<init>", throwable)

        title ="Error"
        (dialogPane.scene.window as Stage).icons += ResourceLookup(ExceptionDialog).image("icon.png")
        headerText = null

        var message: String? = throwable.localizedMessage
        if (message == null) message = throwable.message
        if (message == null) message = throwable.javaClass.simpleName
        contentText = message

        // Create expandable Exception.
        val sw = StringWriter()
        throwable.printStackTrace(PrintWriter(sw))

        // Set expandable Exception into the dialog pane.
        dialogPane.expandableContent = gridpane {
            maxWidth = Double.MAX_VALUE
            label("The exception stacktrace was:") {
                gridpaneConstraints { columnRowIndex(0, 0) }
            }
            textarea(sw.toString()) {
                isEditable = false
                isWrapText = true
                maxWidth = Double.MAX_VALUE
                maxHeight = Double.MAX_VALUE
                gridpaneConstraints {
                    columnRowIndex(0, 1)
                    vhGrow = Priority.ALWAYS
                }
            }
        }
    }

    companion object {
        private var SHOWING = false

        @Synchronized
        fun show(throwable: Throwable) {
            throwable.printStackTrace()

            // show only one instance of the dialog at a time
            if (!SHOWING) {
                SHOWING = true
                ExceptionDialog(throwable).showAndWait()
                SHOWING = false
            }
        }
    }
}
